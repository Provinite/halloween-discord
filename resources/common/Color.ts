import { hexStringToInt } from "./hexStringToInt";

export enum Color {
  Primary = hexStringToInt("EB6123"),
  Error = hexStringToInt("FF0000"),
  Warning = hexStringToInt("FFFF00"),
}
