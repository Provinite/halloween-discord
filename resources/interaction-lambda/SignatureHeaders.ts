export enum SignatureHeaders {
  Signature = "x-Signature-Ed25519",
  Timestamp = "X-Signature-Timestamp",
}
