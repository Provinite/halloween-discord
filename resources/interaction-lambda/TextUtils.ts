import { decode } from "./Base64";

export function base64ToUint8Array(base64Str: string): Uint8Array {
  const binaryString = decode(base64Str);
  const len = binaryString.length;
  const bytes = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes;
}
