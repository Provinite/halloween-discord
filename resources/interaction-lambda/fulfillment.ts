import { SQS } from "aws-sdk";
import { APIApplicationCommandGuildInteraction } from "discord-api-types/v9";
import createHttpError = require("http-errors");
import { FulfillmentMessageBody } from "../common/fulfillment/FulfillmentMessageBody";

export async function sendFulfillmentMessage(
  interaction: APIApplicationCommandGuildInteraction,
  requestTimestamp: number,
): Promise<void> {
  throw new Error("Not implemented.");
  if (!interaction.guild_id) {
    throw new createHttpError[400]("Interaction is missing guild id");
  }
  if (!interaction.member) {
    throw new createHttpError[400]("Interaction is missing member");
  }
  if (!interaction.member.user || !interaction.member.user.id) {
    throw new createHttpError[400]("Interaction is missing member user id");
  }
  if (!interaction.token) {
    throw new createHttpError[400]("Interaction is missing token");
  }
  const sqs = new SQS();
  const messageBody: FulfillmentMessageBody = {
    guildId: interaction.guild_id,
    userId: interaction.member.user.id,
    token: interaction.token,
    interactionId: interaction.id,
    requestTimestamp,
  };
  await sqs
    .sendMessage({
      MessageBody: JSON.stringify(messageBody),
      QueueUrl: "TODO",
      MessageGroupId: interaction.guild_id,
      MessageDeduplicationId: `${interaction.guild_id}+${interaction.member.user.id}+${requestTimestamp}`,
    })
    .promise();
}
