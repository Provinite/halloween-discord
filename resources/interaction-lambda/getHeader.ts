import { APIGatewayProxyEventHeaders } from "aws-lambda";

export function getHeader(
  headers: APIGatewayProxyEventHeaders,
  header: string,
): string | null {
  const result = Object.entries(headers).find(
    ([key]) => key.toLowerCase() === header.toLowerCase(),
  );
  if (!result) {
    return null;
  }
  return result[1] || null;
}
