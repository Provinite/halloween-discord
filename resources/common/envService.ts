import { hexToUint8Array } from "../interaction-lambda/text/TextUtils";

const cache: Record<string, any> = {};

export const envService = {
  getDiscordPublicKey: requiredEnvGetter("DISCORD_PUBLIC_KEY", hexToUint8Array),
  getCommandLambdaArn: requiredEnvGetter("COMMAND_LAMBDA_ARN"),
  getDiscordClientSecret: requiredEnvGetter("DISCORD_CLIENT_SECRET"),
  getDiscordApplicationId: requiredEnvGetter("DISCORD_APPLICATION_ID"),
  getFulfillmentQueueUrl: requiredEnvGetter("FULFILLMENT_QUEUE_URL"),
  getDiscordBotToken: requiredEnvGetter("DISCORD_BOT_TOKEN"),
  getImageBucketUrl: requiredEnvGetter("IMAGE_BUCKET_URL", (s) => {
    if (!s.endsWith("/")) {
      s = s + "/";
    }
    return s;
  }),
};

function requiredEnvGetter(envKey: string): () => string;
function requiredEnvGetter<T = string>(
  envKey: string,
  transform: (s: string) => T,
): () => T;
function requiredEnvGetter<T = string>(
  envKey: string,
  transform?: (s: string) => T,
): () => T {
  return function () {
    const existingResult = cache[envKey];
    if (existingResult) {
      return existingResult;
    }
    let envVar: any = process.env[envKey];
    if (!envVar) {
      throw new Error(`Missing: process.env.${envKey}`);
    }
    if (transform) {
      envVar = transform(envVar);
    }
    cache[envKey] = envVar;
    return envVar;
  };
}
