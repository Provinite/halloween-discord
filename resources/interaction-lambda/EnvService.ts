import { base64ToUint8Array } from "./TextUtils";

export const envService = {
  getDiscordPublicKey(): Uint8Array {
    const base64Key = process.env.DISCORD_PUBLIC_KEY;
    if (!base64Key) {
      throw new Error("Missing: process.env.DISCORD_PUBLIC_KEY");
    }
    return base64ToUint8Array(base64Key);
  },
};
