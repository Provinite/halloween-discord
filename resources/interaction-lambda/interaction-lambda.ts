import { APIGatewayProxyHandler, APIGatewayProxyResult } from "aws-lambda";
import { sign } from "tweetnacl";
import { envService } from "./EnvService";
import { getHeader } from "./getHeader";
import { SignatureHeaders } from "./SignatureHeaders";

export const handler: APIGatewayProxyHandler = async (
  event,
): Promise<APIGatewayProxyResult> => {
  const signature = getHeader(event.headers, SignatureHeaders.Signature);
  const timestamp = getHeader(event.headers, SignatureHeaders.Timestamp);
  const rawBody = event.body || "";

  // Verify signature
  if (!signature || !timestamp) {
    return {
      statusCode: 401,
      body: JSON.stringify({
        error: {
          message:
            'Invalid request, missing signature elements. Requests to this API must be signed with the "X-Signature-Ed25519" and "X-Signature-Timestamp" headers',
        },
      }),
    };
  }

  if (!isVerified(timestamp, rawBody, signature)) {
    return {
      statusCode: 401,
      body: JSON.stringify({
        error: {
          message: "Invalid request, invalid request signature",
        },
      }),
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let body: any = void 0;
  try {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    body = JSON.parse(event.body!);
  } catch (err) {
    body = void 0;
  }
  if (!body) {
    return {
      statusCode: 400,
      body: JSON.stringify({
        error: {
          message: "Invalid request body",
        },
      }),
    };
  }

  // ACK pings
  if (Object.prototype.hasOwnProperty.call(body, "type")) {
    if (body.type === 1) {
      return {
        statusCode: 200,
        body: JSON.stringify(
          {
            type: 1,
          },
          null,
          2,
        ),
      };
    } else {
      return {
        statusCode: 400,
        body: JSON.stringify(
          {
            error: {
              message: "Bad request",
            },
          },
          null,
          2,
        ),
      };
    }
  }

  // Actual interaction handling
  return {
    statusCode: 200,
    body: JSON.stringify({
      type: 4,
      data: {
        tts: false,
        content: "Congrats on sending your command",
        embeds: [],
        allowed_mentions: { parse: [] },
      },
    }),
  };
};

function isVerified(timestamp: string, body: string, signature: string) {
  try {
    return sign.detached.verify(
      Buffer.from(timestamp + body),
      Buffer.from(signature, "hex"),
      envService.getDiscordPublicKey(),
    );
  } catch (err) {
    return false;
  }
}
