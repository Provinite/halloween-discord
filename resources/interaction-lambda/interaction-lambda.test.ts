import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
  Context,
} from "aws-lambda";
import nacl = require("tweetnacl");
import { envService } from "../common/envService";
import { handler } from "./interaction-lambda";
import { SignatureHeaders } from "./SignatureHeaders";

describe("lambda:interaction", () => {
  let publicKey: Uint8Array;
  let secretKey: Uint8Array;
  beforeEach(() => {
    const keyPair = nacl.sign.keyPair();
    publicKey = keyPair.publicKey;
    secretKey = keyPair.secretKey;

    jest.spyOn(envService, "getDiscordPublicKey").mockReturnValue(publicKey);
  });

  describe("signature verification", () => {
    it("errors if signature is not provided", async () => {
      const result = await callHandler({
        headers: {
          [SignatureHeaders.Timestamp]: Date(),
        },
        body: JSON.stringify({
          type: 1,
        }),
      });

      expect(result.statusCode).toBe(401);
      expect(JSON.parse(result.body).error.message).toMatchInlineSnapshot(
        `"Invalid request, missing signature elements. Requests to this API must be signed with the \\"X-Signature-Ed25519\\" and \\"X-Signature-Timestamp\\" headers"`,
      );

      expect(result).toMatchSnapshot();
    });
    it("errors if timestamp is not provided", async () => {
      const result = await callHandler({
        headers: {
          [SignatureHeaders.Signature]: "something that doesn't matter",
        },
        body: JSON.stringify({
          type: 1,
        }),
      });

      expect(result.statusCode).toBe(401);
      expect(JSON.parse(result.body).error.message).toMatchInlineSnapshot(
        `"Invalid request, missing signature elements. Requests to this API must be signed with the \\"X-Signature-Ed25519\\" and \\"X-Signature-Timestamp\\" headers"`,
      );

      expect(result).toMatchSnapshot();
    });
    it("errors if the signature is invalid", async () => {
      const timestamp = Date();

      // sign a cool normal body
      const signedBody = JSON.stringify({
        type: 1,
      });

      // send through a #fakehacker body
      const maliciousBody = JSON.stringify({
        action: "Uninstall discord",
        hacker: "lol",
      });

      const event: Partial<APIGatewayProxyEvent> = {
        headers: {
          [SignatureHeaders.Timestamp]: timestamp,
        },
        body: maliciousBody,
      };

      const signature = nacl.sign.detached(
        Buffer.from(timestamp + signedBody),
        secretKey,
      );
      event.headers![SignatureHeaders.Signature] =
        Buffer.from(signature).toString("hex");
      const result = await callHandler(event);

      expect(result.statusCode).toBe(401);
      expect(JSON.parse(result.body).error.message).toMatchInlineSnapshot(
        `"Invalid request, invalid request signature"`,
      );
      expect(result).toMatchSnapshot();
    });
    it("does not error on properly signed requests", async () => {
      const timestamp = Date();
      const body = JSON.stringify({
        type: 1,
      });
      const event: Partial<APIGatewayProxyEvent> = {
        headers: {
          [SignatureHeaders.Timestamp]: timestamp,
        },
        body,
      };

      const signature = nacl.sign.detached(
        Buffer.from(timestamp + body),
        secretKey,
      );
      event.headers![SignatureHeaders.Signature] =
        Buffer.from(signature).toString("hex");
      const result = await callHandler(event);

      expect(result.statusCode).toBe(200);
    });
  });

  it("must ACK a PING", async () => {
    const result = await callHandlerSigned(
      {
        body: JSON.stringify({ type: 1 }),
      },
      secretKey,
    );

    const bodyObj = JSON.parse(result.body);
    expect(bodyObj.type).toBe(1);
    expect(result.statusCode).toBe(200);

    expect(result).toMatchSnapshot();
  });
});

async function callHandlerSigned(
  event: Partial<APIGatewayProxyEvent> = {},
  secretKey: Uint8Array,
) {
  const timestamp = Date();
  event.headers = event.headers || {};
  event.headers[SignatureHeaders.Timestamp] = timestamp;
  const signature = nacl.sign.detached(
    Buffer.from(timestamp + (event.body || "")),
    secretKey,
  );

  event.headers[SignatureHeaders.Signature] =
    Buffer.from(signature).toString("hex");

  return callHandler(event);
}

async function callHandler(
  event: Partial<APIGatewayProxyEvent> = {},
): Promise<APIGatewayProxyResult> {
  const result = await handler(gatewayEvent(event));

  expect(result).toBeTruthy();
  expect(typeof result).toBe("object");
  expect(result).toHaveProperty("statusCode");
  expect(result).toHaveProperty("body");
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return result!;
}

export function gatewayEvent(
  evt: Partial<APIGatewayProxyEvent> = {},
): APIGatewayProxyEvent {
  return evt as APIGatewayProxyEvent;
}

export function lambdaContext(ctx: Partial<Context> = {}): Context {
  return ctx as Context;
}
