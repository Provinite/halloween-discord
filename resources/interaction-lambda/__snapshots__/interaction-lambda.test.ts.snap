// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambda:interaction must ACK a PING 1`] = `
Object {
  "body": "{
  \\"type\\": 1
}",
  "statusCode": 200,
}
`;

exports[`lambda:interaction signature verification errors if signature is not provided 2`] = `
Object {
  "body": "{\\"error\\":{\\"message\\":\\"Invalid request, missing signature elements. Requests to this API must be signed with the \\\\\\"X-Signature-Ed25519\\\\\\" and \\\\\\"X-Signature-Timestamp\\\\\\" headers\\"}}",
  "statusCode": 401,
}
`;

exports[`lambda:interaction signature verification errors if the signature is invalid 2`] = `
Object {
  "body": "{\\"error\\":{\\"message\\":\\"Invalid request, invalid request signature\\"}}",
  "statusCode": 401,
}
`;

exports[`lambda:interaction signature verification errors if timestamp is not provided 2`] = `
Object {
  "body": "{\\"error\\":{\\"message\\":\\"Invalid request, missing signature elements. Requests to this API must be signed with the \\\\\\"X-Signature-Ed25519\\\\\\" and \\\\\\"X-Signature-Timestamp\\\\\\" headers\\"}}",
  "statusCode": 401,
}
`;
